Use enum as key in IDL

I met some problems while using enum as key in IDL. It seems that enum can't be key in the latest CycloneDDS.

How to reproduce:

1. Get the latest CycloneDDS
```bash
git clone https://github.com/eclipse-cyclonedds/cyclonedds.git
```

2. Modify the example code Helloworld IDL `examples/helloworld/HelloWorldData.idl` like the following:
```
module HelloWorldData
{
  enum TYPE {
    TYPE1,
    TYPE2
  };
  struct Msg
  {
    long userID;
    string message;
    TYPE type;
  };
  #pragma keylist Msg type
}; 
```

3. Build the code
```bash
cmake -DBUILD_EXAMPLES=ON -Bbuild -H. && cmake --build build
```

4. Then it'll pop up the following error message.
```
/home/chenying/cyclonedds/examples/helloworld/HelloWorldData.idl:13:21: Invalid key 'type' type in keylist directive
examples/helloworld/CMakeFiles/HelloWorldData_lib_generate.dir/build.make:62: recipe for target 'examples/helloworld/HelloWorldData.c' failed
make[2]: *** [examples/helloworld/HelloWorldData.c] Error 1
CMakeFiles/Makefile2:1125: recipe for target 'examples/helloworld/CMakeFiles/HelloWorldData_lib_generate.dir/all' failed
make[1]: *** [examples/helloworld/CMakeFiles/HelloWorldData_lib_generate.dir/all] Error 2
Makefile:151: recipe for target 'all' failed
make: *** [all] Error 2
```

However, the problem doesn't exist in version 0.7.0. I'm not sure whether it's legal to use enum as key in IDL.
If it's ok to use, maybe the problem is related to the if-condition here:
https://github.com/eclipse-cyclonedds/cyclonedds/blob/46eda2d7786b55ca9704470b58c15cff3b93621f/src/idl/src/directive.c#L339

--- a/src/idl/src/directive.c
+++ b/src/idl/src/directive.c
@@ -324,6 +324,7 @@ push_keylist(idl_pstate_t *pstate, struct keylist *dir)
 
   for (size_t i=0; dir->keys && dir->keys[i]; i++) {
     idl_key_t *key = NULL;
+    idl_mask_t mask = IDL_BASE_TYPE | IDL_ENUM | IDL_STRING;
     const idl_declarator_t *declarator;
     const idl_type_spec_t *type_spec;
 
@@ -335,10 +336,15 @@ push_keylist(idl_pstate_t *pstate, struct keylist *dir)
     declarator = (const idl_declarator_t *)declaration->node;
     assert(idl_is_declarator(declarator));
     type_spec = idl_type_spec(declarator);
+    /* until DDS-XTypes is fully implemented, base types, enums, arrays of the
+       aforementioned and strings are allowed to be used in keys */
+    type_spec = idl_unalias(type_spec, 0u);
+    if (idl_is_array(type_spec))
+      mask &= (idl_mask_t)~IDL_STRING;
     type_spec = idl_unalias(type_spec, IDL_UNALIAS_IGNORE_ARRAY);
-    if (!(idl_is_base_type(type_spec) || idl_is_string(type_spec))) {
+    if (!(idl_mask(type_spec) & mask)) {
       idl_error(pstate, idl_location(dir->keys[i]),
-        "Invalid key '%s' type in keylist directive", dir->keys[i]->identifier);
+        "Invalid key '%s' in keylist directive", dir->keys[i]->identifier);
       return IDL_RETCODE_SEMANTIC_ERROR;
     }
 

--- a/src/idl/tests/pragma.c
+++ b/src/idl/tests/pragma.c
@@ -18,18 +18,29 @@
 
 #include "CUnit/Test.h"
 
+static idl_retcode_t parse_string(const char *str, idl_pstate_t **pstatep)
+{
+  idl_retcode_t ret;
+  idl_pstate_t *pstate = NULL;
+  if ((ret = idl_create_pstate(0u, NULL, &pstate)))
+    return ret;
+  assert(pstate);
+  ret = idl_parse_string(pstate, str);
+  if (ret == IDL_RETCODE_OK)
+    *pstatep = pstate;
+  else
+    idl_delete_pstate(pstate);
+  return ret;
+}
+
 static void test_bad_use(const char *str)
 {
   idl_retcode_t ret;
   idl_pstate_t *pstate = NULL;
 
-  ret = idl_create_pstate(0u, NULL, &pstate);
-  CU_ASSERT_EQUAL_FATAL(ret, IDL_RETCODE_OK);
-  CU_ASSERT_PTR_NOT_NULL_FATAL(pstate);
-  assert(pstate);
-  ret = idl_parse_string(pstate, str);
+  ret = parse_string(str, &pstate);
   CU_ASSERT_EQUAL(ret, IDL_RETCODE_SEMANTIC_ERROR);
-  CU_ASSERT_PTR_NULL(pstate->root);
+  CU_ASSERT_PTR_NULL(pstate);
   idl_delete_pstate(pstate);
 }
 
@@ -55,36 +66,63 @@ CU_Test(idl_pragma, keylist_bad_use)
     test_bad_use(*ptr);
 }
 
-static idl_retcode_t parse_string(const char *str, idl_pstate_t **pstatep)
+static void test_keylist(const char *str, idl_retcode_t exp)
 {
   idl_retcode_t ret;
   idl_pstate_t *pstate = NULL;
-  if ((ret = idl_create_pstate(0u, NULL, &pstate)))
-    return ret;
-  assert(pstate);
-  ret = idl_parse_string(pstate, str);
-  if (ret == IDL_RETCODE_OK)
-    *pstatep = pstate;
-  else
-    idl_delete_pstate(pstate);
-  return ret;
+
+  ret = parse_string(str, &pstate);
+  CU_ASSERT_EQUAL(ret, exp);
+  if (ret == IDL_RETCODE_OK) {
+    CU_ASSERT_PTR_NOT_NULL_FATAL(pstate);
+    CU_ASSERT_PTR_NOT_NULL_FATAL(pstate->root);
+    const void *node = pstate->root;
+    const idl_keylist_t *keylist = NULL;
+    while (idl_is_enum(node))
+      node = idl_next(node);
+    CU_ASSERT(idl_is_struct(node));
+    if (node && idl_is_struct(node))
+      keylist = ((const idl_struct_t *)node)->keylist;
+    CU_ASSERT_PTR_NOT_NULL(keylist);
+  }
+  idl_delete_pstate(pstate);
 }
 
 CU_Test(idl_pragma, keylist)
 {
-  idl_retcode_t ret;
-  idl_pstate_t *pstate = NULL;
-  idl_struct_t *s1;
-  const char str[] = "struct s1 { char c; };\n"
-                     "#pragma keylist s1 c";
+#define TEST(type) "struct s { " type " m %s; };\n" \
+                   "#pragma keylist s m\n"
 
-  ret = parse_string(str, &pstate);
-  CU_ASSERT_EQUAL_FATAL(ret, IDL_RETCODE_OK);
-  assert(pstate);
-  s1 = (idl_struct_t *)pstate->root;
-  CU_ASSERT_FATAL(idl_is_struct(s1));
-  CU_ASSERT((idl_mask(s1->keylist) & IDL_KEYLIST) != 0);
-  idl_delete_pstate(pstate);
+  const char *fmts[] = {
+    /* base types */
+    TEST("boolean"),
+    TEST("char"),
+    TEST("octet"),
+    TEST("short"),
+    TEST("unsigned short"),
+    TEST("long"),
+    TEST("unsigned long"),
+    TEST("long long"),
+    TEST("unsigned long long"),
+    /* enums */
+    "enum abc { A, B };\n" TEST("abc"),
+    /* strings */
+    TEST("string")
+  };
+#undef TEST
+
+  for (size_t i=0, n=(sizeof(fmts)/sizeof(fmts[0])); i < n; i++) {
+    char buf[128];
+
+    /* type */
+    (void)snprintf(buf, sizeof(buf), fmts[i], "");
+    test_keylist(buf, IDL_RETCODE_OK);
+    if (i == n - 1) /* skip array test for strings */
+      continue;
+    /* array of type */
+    (void)snprintf(buf, sizeof(buf), fmts[i], "[2]");
+    test_keylist(buf, IDL_RETCODE_OK);
+  }
 }
 
 CU_Test(idl_pragma, keylist_nested_key)

